provisioning.go

```diff
@@ -0,0 +1,370 @@
+	AttrValidSRK ProvisionStatusAttributes = 1 << iota
+
+	// AttrValidEK indicates that the TPM contains a valid endorsement key at the expected location. On a TPMConnection created
+	// with SecureConnectToDefaultTPM, it means that the TPM contains the key associated with the verified endorsment certificate.
+	// On a TPMConnection created with ConnectToDefaultTPM, it means that the TPM contains a valid primary key with the expected
+	// properties at the expected location, but does not mean that the object was created with the the same template that
+	// ProvisionTPM uses, and is no guarantee that a call to ProvisionTPM wouldn't result in a different key being created.
+	AttrValidEK
+
+	AttrDAParamsOK         // The dictionary attack lockout parameters are configured correctly.
+	AttrOwnerClearDisabled // The ability to clear the TPM with owner authorization is disabled.
+
+	// AttrLockoutAuthSet indicates that the lockout hierarchy has an authorization value defined. This
+	// doesn't necessarily mean that the authorization value is the same one that was originally provided
+	// to ProvisionTPM - it could have been changed outside of our control.
+	AttrLockoutAuthSet
+
+	AttrLockNVIndex // The TPM has a valid NV index used for locking access to keys sealed with SealKeyToTPM
+)
+
+const (
+	// ProvisionModeClear specifies that the TPM should be fully provisioned after clearing it.
+	ProvisionModeClear ProvisionMode = iota
+
+	// ProvisionModeWithoutLockout specifies that the TPM should be refreshed without performing
+	// operations that require knowledge of the lockout hierarchy authorization value.
+	ProvisionModeWithoutLockout
+
+	// ProvisionModeFull specifies that the TPM should be fully provisioned without clearing it.
+	ProvisionModeFull
+)
+
+var (
+	// srkTemplate is the default RSA2048 SRK template, see section 7.5.1 of "TCG TPM v2.0 Provisioning Guidance", version 1.0,
+	// revision 1.0, 15 March 2017.
+	srkTemplate = tpm2.Public{
+		Type:    tpm2.ObjectTypeRSA,
+		NameAlg: tpm2.HashAlgorithmSHA256,
+		Attrs: tpm2.AttrFixedTPM | tpm2.AttrFixedParent | tpm2.AttrSensitiveDataOrigin | tpm2.AttrUserWithAuth | tpm2.AttrNoDA |
+			tpm2.AttrRestricted | tpm2.AttrDecrypt,
+		Params: tpm2.PublicParamsU{
+			Data: &tpm2.RSAParams{
+				Symmetric: tpm2.SymDefObject{
+					Algorithm: tpm2.SymObjectAlgorithmAES,
+					KeyBits:   tpm2.SymKeyBitsU{Data: uint16(128)},
+					Mode:      tpm2.SymModeU{Data: tpm2.SymModeCFB}},
+				Scheme:   tpm2.RSAScheme{Scheme: tpm2.RSASchemeNull},
+				KeyBits:  2048,
+				Exponent: 0}},
+		Unique: tpm2.PublicIDU{Data: make(tpm2.PublicKeyRSA, 256)}}
+)
+
+// ProvisionTPM prepares the TPM associated with the tpm parameter for full disk encryption. The mode parameter specifies the
+// behaviour of this function.
+//
+// If mode is not ProvisionModeWithoutLockout, this function performs operations that require knowledge of the lockout hierarchy
```

Comment:
I would move this later together with the explanation of what we try to do in ProvisionModeWithoutLockout.